packages/frontend/src/redux/slices/ledger/index.js

```diff
@@ -1,14 +1,192 @@
+
+            unset(state, ['txSigned']);
+            set(state, ['modal', 'show'], !signInWithLedgerStatus && payload.show);
+            set(state, ['modal', 'action'], payload.action);
+            set(state, ['modal', 'textId'], 'ledgerSignTxModal.DEFAULT');
+        },
+        hideLedgerModal(state, { payload, ready, error }) {
+            set(state, ['modal'], {});
+            unset(state, ['txSigned']);
+        },
+    },
+    extraReducers: ((builder) => {
+        // getLedgerAccountIds
+        builder.addCase(getLedgerAccountIds.pending, (state) => {
+            set(state, ['signInWithLedgerStatus'], LEDGER_MODAL_STATUS.CONFIRM_PUBLIC_KEY);
+        });
+        builder.addCase(getLedgerAccountIds.fulfilled, (state, { payload }) => {
+            unset(state, ['txSigned']);
+            set(state, ['signInWithLedgerStatus'], LEDGER_MODAL_STATUS.CONFIRM_ACCOUNTS);
+            payload.forEach(accountId => 
+                set(state, ['signInWithLedger', accountId, 'status'], 'waiting')
+            );
+        });
+        builder.addCase(getLedgerAccountIds.rejected, (state, { error }) => {
+            const noAccounts = error.message === 'No accounts were found.' && !HIDE_SIGN_IN_WITH_LEDGER_ENTER_ACCOUNT_ID_MODAL;
+
+            set(state, ['signInWithLedgerStatus'], noAccounts ? LEDGER_MODAL_STATUS.ENTER_ACCOUNTID : undefined);
+            unset(state, ['signInWithLedger']);
+            unset(state, ['txSigned']);
+        });
+        // addLedgerAccountId
+        builder.addCase(addLedgerAccountId.pending, (state, { payload, meta: { arg: { accountId } } }) => {
+            set(state, ['signInWithLedgerStatus'], LEDGER_MODAL_STATUS.CONFIRM_ACCOUNTS);
+            set(state, ['signInWithLedger', accountId, 'status'], 'confirm');
+        });
+        builder.addCase(addLedgerAccountId.fulfilled, (state, { payload, meta: { arg: { accountId } } }) => {
+            set(state, ['signInWithLedgerStatus'], LEDGER_MODAL_STATUS.CONFIRM_ACCOUNTS);
+            set(state, ['signInWithLedger', accountId, 'status'], 'success');
+        });
+        builder.addCase(addLedgerAccountId.rejected, (state, { error, meta: { arg: { accountId } } }) => {
+            set(state, ['signInWithLedgerStatus'], LEDGER_MODAL_STATUS.CONFIRM_ACCOUNTS);
+
+            const transportError = error?.name === 'TransportStatusError';
+            set(state, ['signInWithLedger', accountId, 'status'], transportError ? 'rejected' : 'error');
+        });
+        // refreshAccountOwner
+        builder.addCase(refreshAccountOwner.fulfilled, (state, { payload }) => {
+            set(state, ['hasLedger'], payload.ledger.hasLedger);
+            set(state, ['ledgerKey'], payload.ledger.ledgerKey);
+        });
+        // matcher to handle closing modal automatically
+        builder.addMatcher(
+            ({ type, ready, error }) => ready || error || type.endsWith('/rejected') || type.endsWith('/fulfilled'),
```

Comment:
Curious about why we need the fallback to check the type here - can `ready` and `error` both be `false` when the promise has fulfilled/rejected? Does that mean that `ready` and `error` are strictly synchronous?

Reply:
Yes, that's because `readyStatePromise` async actions will have ready and error param and `createAsyncThunk` actions will use `rejected` and `fulfilled` types of actions only. For now, we need to check both scenarios.

This might be also simplified but only after we use `redux-toolkit` reducers only.